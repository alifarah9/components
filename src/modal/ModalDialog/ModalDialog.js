import React, { Component } from 'react';
import classNames from 'classnames';

import { dialogPropTypes, dialogDefaultProps } from '../props';
import { addModalOpenBodyClass, removeModalOpenBodyClass } from './DOMOperations';
import './ModalDialog.less';

export default class ModalDialog extends Component {
  static propTypes = dialogPropTypes;
  static defaultProps = dialogDefaultProps;

  componentDidMount() {
    addModalOpenBodyClass();
    if (this.modalDialog) {
      this.modalDialog.focus();
      document.addEventListener('keydown', this.onEscape);
    }
  }

  componentWillUnmount() {
    removeModalOpenBodyClass();
    document.removeEventListener('keydown', this.onEscape);
  }

  onEscape = event => {
    if (event.key === 'Escape' && this.props.onClose) {
      this.props.onClose(event);
    }
  };

  render() {
    const { title, body, footer, onClose, role, className, ...otherProps } = this.props;

    const dialogBaseClass = 'modal-dialog';
    const classes = classNames(dialogBaseClass, {
      [`modal-${this.props.size}`]: this.props.size,
    });

    return (
      <div
        className={`modal fade ${className}`}
        style={{ display: 'block' }}
        tabIndex="-1"
        role={role}
        aria-modal
        ref={dialog => {
          this.modalDialog = dialog;
        }}
        {...otherProps}
      >
        <div className={classes}>
          <div className="modal-content">
            <div className={classNames('modal-header', { 'no-divider': !title })}>
              <button type="button" onClick={onClose} className="close" aria-label="close">
                <span aria-hidden="true">&times;</span>
              </button>
              <h4 className="modal-title">{title}</h4>
            </div>
            <div className={classNames('modal-body', { 'modal-body--no-title': !title })}>
              {body}
            </div>
            {footer && <div className="modal-footer">{footer}</div>}
          </div>
        </div>
      </div>
    );
  }
}
