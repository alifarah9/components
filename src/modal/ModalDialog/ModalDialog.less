/*
 We will not need this in very near future as we would want all our modals,
 to be full-screen on mobiles. This code will then move to our common CSS
 library the @transferwise/bootstrap repo.
 */
@import '~bootstrap/less/variables';

@media only screen and (max-width: @screen-md) {
  .modal-dialog {
    height: 100%;
    max-width: 100%;
    width: 100%;
    margin: 0;
    padding: 0;
  }

  .modal-content {
    position: relative;
    display: flex;
    flex-direction: column;
    pointer-events: auto;
    background-clip: padding-box;
    outline: 0;
    height: 100%;
    border-radius: 0;
  }

  .modal-header {
    justify-content: flex-start;
    min-height: 50px;
    line-height: 50px;
    padding: 16px;
  }

  .modal-body {
    overflow: auto;
    position: relative;
    flex: 1 1 auto;
    padding: 15px;
    flex-shrink: 3;
  }

  .modal-footer {
    margin: 0;
    padding: 16px;
    .btn {
      display: block;
      width: 100%;
    }
  }
}

@media (min-width: @screen-sm-min) {
  .modal-xl {
    width: auto;
    margin: (@padding-base-vertical * 4) @grid-gutter-width;
  }
}

@media (min-width: @screen-lg-min) {
  .modal-xl {
      width: 1000px;
      margin: (@padding-base-vertical * 4) auto;
  }
}

.modal-footer {
  .btn {
    display: inline-block;
  }
}
