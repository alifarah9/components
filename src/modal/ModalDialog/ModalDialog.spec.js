import React, { Fragment } from 'react';
import { shallow, mount } from 'enzyme';
import { addModalOpenBodyClass, removeModalOpenBodyClass } from './DOMOperations';

import ModalDialog from '.';

jest.mock('./DOMOperations');

describe('ModalDialog', () => {
  let component;
  beforeEach(() => {
    component = shallow(<ModalDialog title="Some title" body="Some body" onClose={jest.fn()} />);
  });

  // Mainly snapshot tests with contents being checked in snapshots.
  describe('renders', () => {
    it('default dialog', () => {
      expect(component).toMatchSnapshot();
    });

    it('with small size when passed down the props', () => {
      component.setProps({ size: 'sm' });
      expect(component).toMatchSnapshot();
    });

    it('with medium sized when passed down the props', () => {
      component.setProps({ size: 'md' });
      expect(component).toMatchSnapshot();
    });

    it('with large sized when passed down the props', () => {
      component.setProps({ size: 'lg' });
      expect(component).toMatchSnapshot();
    });

    it('with title', () => {
      component.setProps({
        title: 'A title',
      });

      expect(component).toMatchSnapshot();
    });

    it('with body content', () => {
      component.setProps({
        body: <p>Some paragraph text</p>,
      });

      expect(component).toMatchSnapshot();
    });

    it('with no footer when `footer` not specified as prop', () => {
      component.setProps({
        footer: null,
      });

      expect(component).toMatchSnapshot();
    });

    it('with footer content when `footer` specified as prop', () => {
      component.setProps({
        footer: <button>Some button</button>,
      });

      expect(component).toMatchSnapshot();
    });
  });

  // Mainly behaviour tests
  describe('behaviourally', () => {
    let originalAddEventListener;
    let documentEventCallbacks;
    beforeEach(() => {
      originalAddEventListener = global.document.addEventListener;
      documentEventCallbacks = {};
      global.document.addEventListener = jest.fn((name, cb) => {
        documentEventCallbacks[name] = cb;
      });
    });

    afterEach(() => {
      global.document.addEventListener = originalAddEventListener;
    });
    it('calls close handler on close button click', () => {
      const onClose = jest.fn();
      component.setProps({ onClose });

      expect(onClose).not.toBeCalled();
      clickCloseButton();
      expect(onClose).toBeCalled();
    });

    it('closes on `esc` keypress', () => {
      jest.clearAllMocks();
      const onClose = jest.fn();
      component = mount(<ModalDialog title="Some title" body="Some body" onClose={onClose} />);

      expect(onClose).not.toBeCalled();
      pressEscapeOnComponent();
      expect(onClose).toBeCalled();
    });

    it('executes modal open operations on mount', () => {
      jest.clearAllMocks();

      expect(addModalOpenBodyClass).not.toBeCalled();
      component = shallow(<ModalDialog title="Some title" body="Some body" onClose={jest.fn()} />);
      expect(addModalOpenBodyClass).toBeCalled();
    });

    it('executes modal close operations on unmount', () => {
      expect(removeModalOpenBodyClass).not.toBeCalled();
      component.unmount();
      expect(removeModalOpenBodyClass).toBeCalled();
    });

    it('removes event listener on unmount', () => {
      jest.clearAllMocks();
      const onClose = jest.fn();
      component = mount(<ModalDialog title="Some title" body="Some body" onClose={onClose} />);
      jest.spyOn(component.instance(), 'onEscape');
      const listener = component.instance().onEscape;
      component.unmount();
      pressEscapeOnBody();
      expect(listener).not.toBeCalled();
    });

    function clickCloseButton() {
      component.find('.close').simulate('click');
    }

    function pressEscapeOnComponent() {
      component.simulate('keyDown');
      documentEventCallbacks.keydown({ key: 'Escape' });
      component.update();
    }

    function pressEscapeOnBody() {
      documentEventCallbacks.keydown({ key: 'Escape', target: document });
      component.update();
    }
  });
});
