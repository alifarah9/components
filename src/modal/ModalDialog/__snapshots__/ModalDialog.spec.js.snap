// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalDialog renders default dialog 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with \`modal-body--no-title\` class if title is not provided 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with a custom class 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with body content 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with footer content when \`footer\` specified as prop 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with header divider if title is provided 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with large sized when passed down the props 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with medium sized when passed down the props 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with no footer when \`footer\` not specified as prop 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with small size when passed down the props 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with some random other props 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders with title 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders without \`modal-body--no-title\` class if title is provided 1`] = `ShallowWrapper {}`;

exports[`ModalDialog renders without header divider if title is not provided 1`] = `ShallowWrapper {}`;
